include/clap/ext/draft/context-menu.h

```diff
@@ -0,0 +1,121 @@
+#pragma once
+
+#include "../../plugin.h"
+
+// This extension lets the host and plugin exchange menu items and let the plugin ask the host to
+// show its context menu.
+
+static CLAP_CONSTEXPR const char CLAP_EXT_CONTEXT_MENU[] = "clap.context-menu.draft/0";
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+// There can be different target kind for a context menu
+enum {
+   CLAP_CONTEXT_MENU_TARGET_KIND_GLOBAL,
+   CLAP_CONTEXT_MENU_TARGET_KIND_PARAM,
+};
+
+// Describes the context menu target
+typedef struct clap_context_menu_target
+{
+   uint32_t kind;
+   clap_id id;
+} clap_context_menu_target_t;
+
+// Flags for the context menu accelerator
+enum {
+   CLAP_CONTEXT_MENU_ACCELERATOR_MODIFIER_SHIFT = (1 << 0),
+   CLAP_CONTEXT_MENU_ACCELERATOR_MODIFIER_CONTROL = (1 << 1),
+   CLAP_CONTEXT_MENU_ACCELERATOR_MODIFIER_ALT = (1 << 2),
+
+   // Windows key or Command key
+   CLAP_CONTEXT_MENU_ACCELERATOR_MODIFIER_META1 = (1 << 3),
+};
+
+// Describes a single key strike with its modifiers
+typedef struct clap_context_menu_accelerator_key {
+   uint32_t flags;
+   uint32_t key; // ASCII key
+} clap_context_menu_accelerator_key_t;
+
+// Describes an accelerator which can consist of a key strike combo
+typedef struct clap_context_menu_accelerator {
+   uint32_t combo_size;
+   clap_context_menu_accelerator_key_t keys[4];
```

Comment:
Why keys[4]? Just one would do I think.

Reply:
Key chords. But, I still don't think any of this is useful or even usable for an extension like this. An `accelerator` parameter that accepts a single character string should suffice.

Reply:
I guess the term chords for this is not very widely used outside of Emacs. What I'm referring to here is a key binding with one or more leader keys. So for instance `Ctrl+C` followed by `Alt+G` followed by `Ctrl+C` followed by `d`, or `C-c M-g C-c d` for short. But, I don't know why context menus would need keybindings like this (and again, keybindings more complex than single letter accelerators aren't supported in most toolkits).

Reply:
Obligatory xkcd: https://xkcd.com/2583/

I also think it is unlikely a plugin and host could negotiate non overlapping keyboard accelerators. Like if bitwig tells me alt 3 does something and I use it for something else I will ignore bitwig. And if I don’t have keybindgd I will ignore bitwig. The accelerator display is useful perhaps but are you expecting the plugin will actually process the key events!

Reply:
Key chords... learned something new ;-)

I guess a host doesn't have that many shortcuts of its own while a plugin has focus? It could work, but it's probably overkill indeed.

Reply:
1. Idea A: Maybe there should be an extra extension for plugin keyboard accelerators?
I mean, it always has been problematic with overlapping host/plugin accelerators. So if they would communicate about that with each other, they could maybe find a better compromise? Like the plugin requests a set of accelerators or the host can tell the plugin which accelerators are forbidden. Something like this. So I arrive at...

2. Idea B: Or maybe keyboard accelerator concepts would better fit into a "commands"(&settings) extension where the plugin tells the host its set of commands/toggle settings/enum setting, so that the host could set up a plugin-specific key command editor (where it can prevent the user from assigning duplicate accelerators).
(Another argument for this approach would also be that it is unlikely that there are key accelerators for parameter-specific commands/settings but rather only for global commands.)
Also a separate commands/settings extension would allow the host to MIDI-learn plugin commands which would be awesome (but would not really fit into a context menu extension)

Reply:
Ah yes, my most important point: All in all I think that we maybe even could leave the accelerators completely out here... Or just make it a simple character string as others already proposed here (even if the plugin cannot guarantee that the accelerator will actually work)

Reply:
Re: 1), accelerators are allowed to overlap in pretty much every GUI toolkit. You'd just press the same key to cycle between the options and then hit enter. I personally never use the context menu accelerators at all though.

Reply:
Oh, I think I mixed up terminology here: My entire comments were about shortcuts/key commands, not accelerators

Reply:
In a context menu, there would be key bindings like a simple letter. As example on Windows this is such a thing as "&Cut" where `C` gets underlined, and a keypress to `C` selects the entry. This is called **navigation key** or **hotkey**. They work only when the context menu is open.

Accelerators are usually a keyboard shortcut for a (program global) function in a menu bar above the window.
Here CTRL+C would be the accelerator for "Cut", where the entry is looking like "Cut     CTRL-C".

Since this is a context menu, we talk about hotkeys, right? In this case, modifiers would not be needed, since using a modifier key immediately closes such a menu.

Reply:
Thank you for clarification, @defiantnerd! Ok, then I really talked about "accelerators". :) Really sorry for confusion! :/

Reply:
So, should we remove the accelerator part? It is a context menu, opened with right-click, correct?
A different beast would be adding menu entries to a host provided menu bar.

Reply:
Sorry for the delay, last week has been intense for me and I could not find the time to get back to it.

I'd like to clarify one aspect of this extension:
- accelerator are just here to tell the user how the plugin (or host depending on who provides the menu item) is processing its shortcut, it isn't meant to provide a way to exchange accelerators (or shortcuts) between the host and plugins

Then about simplifying the accelerator, if instead we describe them using a string `"&A"` for `Alt+A` how do you then do `Alt+&` is it `"&&"`?
So we could have:
- `#` for Command/Windows
- `&` for Alt
- `@` for Ctrl
- `!` for Shift
- `,` for separating key chords

`"&#k,!A"` for `Alt+Cmd+k, Shift+A`
`"&##"` for `Alt+Cmd+#`

Is that really better?

Reply:
To me, using a string would then force to have a parser which will then produce something equivalent to what `clap_context_menu_accelerator_key` is. So I'm far from being convinced here, also when your application becomes more complex, you probably have an internal representation of a shortcut which isn't a string itself.

Here is what most will do:
```C
clap_context_menu_accelerator_key_t accel = convert_to_clap(my_accel);
...
```

The current interface is maybe more work at first, but I suppose that in the long run, there will be less issues, because parsing a string is always a source of bugs.